### PR TITLE
Parsec sequestre dump

### LINKED ISSUE #2364: [Sequester] RealmExtract access CLI
MVP:
- CLI only
- dump decrypted workspace file&folder on the disk
- only consider last data (i.e. don't provide history crawling support)

example:
```
$ parsec_sequestre_dump --input my/workspace_export.sqlite --key my/keyfile.key --output ./output
```

Bonus points:
- timestamp based history crawling support

### PR DESCRIPTION
**Pull request by TimeEngineer**
Rename of #2363

### CODE REVIEW

File: oxidation/parsec_cli/parsec_sequestre_dump/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+use clap::Parser;
+use diesel::{Connection, SqliteConnection};
+use std::path::PathBuf;
+
+use crate::model::Data;
+use parsec_api_crypto::SecretKey;
+
+#[derive(Parser, Debug)]
+#[clap(author, version, about, long_about = None)]
+struct Args {
```
Comment (FirelightFlagboy) [inline oxidation/parsec_cli/parsec_sequestre_dump/src/main.rs:18]: 1. add missing help message
2. use [`validator`](https://docs.rs/clap/latest/clap/struct.Arg.html#method.validator) to check if the file exist (for `input` and `key`)

File: oxidation/parsec_cli/parsec_sequestre_dump/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+use clap::Parser;
+use diesel::{Connection, SqliteConnection};
+use std::path::PathBuf;
+
+use crate::model::Data;
+use parsec_api_crypto::SecretKey;
+
+#[derive(Parser, Debug)]
+#[clap(author, version, about, long_about = None)]
```
Comment (FirelightFlagboy) [inline oxidation/parsec_cli/parsec_sequestre_dump/src/main.rs:17]:
```suggestion
#[clap(version, about)]
```
- not forced to set *long about* to `None`
- `Cargo.toml` don't have an `authors` section

File: oxidation/parsec_cli/parsec_sequestre_dump/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,16 @@
+[package]
```
Comment (FirelightFlagboy) [inline oxidation/parsec_cli/parsec_sequestre_dump/Cargo.toml:1]: you may want to add a description for `clap` to use it in the `about` message

File: oxidation/parsec_cli/parsec_sequestre_dump/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+struct Args {
+    /// Sqlite input file (e.g `my/workspace_export.sqlite`)
+    #[clap(short, long)]
+    input: PathBuf,
+
+    /// Secret key (e.g `my/keyfile.key`)
+    #[clap(short, long)]
+    key: String,
+
+    /// Output disk file (e.g `./output`)
```
Comment (FirelightFlagboy) [inline oxidation/parsec_cli/parsec_sequestre_dump/src/main.rs:27]:
```suggestion
    /// Output file (e.g `./output`)
```
did you really mean `disk file` ? just file seems enough ?

**Review by FirelightFlagboy — CHANGES_REQUESTED** [review]
It's expected that the dumped file is encrypted ?

File: oxidation/parsec_cli/parsec_sequestre_dump/src/main.rs (hunk 4 above)
Comment (TimeEngineer) [reply oxidation/parsec_cli/parsec_sequestre_dump/src/main.rs:27]: Done

File: oxidation/parsec_cli/parsec_sequestre_dump/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+    use rstest::rstest;
+    use tests_fixtures::{alice, Device};
+
+    use crate::model::{Data, Workspace};
+
+    #[rstest]
+    fn test_parsec_sequestre_dump(alice: &Device) {
+        use diesel::connection::SimpleConnection;
+
+        let input = "/tmp/parsec_sequestre_dump.sqlite";
```
Comment (touilleMan) [inline oxidation/parsec_cli/parsec_sequestre_dump/src/main.rs:61]: we need a `tmp_path` fixture that provide a temporary folder instead of always writing in the same place (this is bad because it can lead do concurrent issues, especially with multiple tests using that all initialize with the same location due to copy/paste coding)

Comment (touilleMan) [reply oxidation/parsec_cli/parsec_sequestre_dump/src/main.rs:61]: and the `tmp_path` will be great to be put in the `tests_fixtures` crate given it most likely is going to be needed in other plenty of other tests ;-)

File: oxidation/crates/parsec_api_types/tests/test_addr.rs [hunk 6]
```diff
@@ -158,14 +158,6 @@ fn all_addr(testbed: &dyn Testbed) {}
 )]
 fn addr_with_org(testbed: &dyn Testbed) {}
 
-#[template]
```
Comment (touilleMan) [inline oxidation/crates/parsec_api_types/tests/test_addr.rs:161]: This is not related to this PR, why removing ?

File: oxidation/parsec_cli/parsec_sequestre_dump/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+parsec_api_crypto = { path = "../../crates/parsec_api_crypto" }
+parsec_api_types = { path = "../../crates/parsec_api_types" }
+
+diesel = { version = "2.0.0-rc.0", features = ["sqlite", "chrono"] }
+clap = { version = "3.1.18", features = ["derive"] }
+serde = { version = "1.0.127", features = ["derive"] }
```
Comment (touilleMan) [inline oxidation/parsec_cli/parsec_sequestre_dump/Cargo.toml:15]: why serde/serde_json ? deserialization should all be handled by `parsec_api_types`

File: oxidation/parsec_cli/parsec_sequestre_dump/src/main.rs [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+            })
+        );
+
+        workspace.dump(output).unwrap();
+
+        let loaded_data = Workspace::load(output);
+
+        assert_eq!(workspace, loaded_data);
+
+        std::fs::remove_file(input).unwrap();
```
Comment (touilleMan) [inline oxidation/parsec_cli/parsec_sequestre_dump/src/main.rs:173]: Doing this means the files won't be removed if an assert fails, (this is why fixture with there teardown step always executed is so convenient ^^)

File: oxidation/parsec_cli/parsec_sequestre_dump/src/model.rs [hunk 9]
```diff
@@ -0,0 +1,519 @@
...
+    }
+
+    #[cfg(test)]
+    pub fn load(path: &str) -> Self {
+        let data = std::fs::read(path).unwrap();
+        serde_json::from_slice(&data).unwrap()
+    }
+
+    #[cfg(test)]
+    pub fn init_db(conn: &mut SqliteConnection, device: &Device) {
```
Comment (touilleMan) [inline oxidation/parsec_cli/parsec_sequestre_dump/src/model.rs:234]: This is totally related to the test, you should have all this in a `tests` folder instead

File: oxidation/parsec_cli/parsec_sequestre_dump/src/model.rs [hunk 10]
```diff
@@ -0,0 +1,519 @@
...
+
+        let workspace = vlob_atom::table
+            .select(vlob_atom::blob)
+            .filter(vlob_atom::vlob_id.eq(workspace_id))
+            .first::<Vec<u8>>(conn)?;
+
+        Ok(Self(Data::load(conn, key, String::new(), &workspace)?))
+    }
+
+    #[cfg(test)]
```
Comment (touilleMan) [inline oxidation/parsec_cli/parsec_sequestre_dump/src/model.rs:227]: Those should be in a `tests/common.rs` file instead

File: oxidation/parsec_cli/parsec_sequestre_dump/src/model.rs [hunk 11]
```diff
@@ -0,0 +1,519 @@
...
+
+type Time = chrono::DateTime<chrono::Utc>;
+
+#[cfg(test)]
+use crate::schema::{device, realm_role, user_ as user};
+#[cfg(test)]
+use tests_fixtures::Device;
+
+#[derive(Debug, Queryable, Serialize, Deserialize, PartialEq)]
+pub struct Block {
```
Comment (touilleMan) [inline oxidation/parsec_cli/parsec_sequestre_dump/src/model.rs:21]: no User/Device/RealmRole structs ?

File: oxidation/parsec_cli/parsec_sequestre_dump/src/model.rs [hunk 12]
```diff
@@ -0,0 +1,519 @@
...
+use parsec_api_types::{DeviceID, FileManifest, FolderManifest, Manifest, WorkspaceManifest};
+
+type Time = chrono::DateTime<chrono::Utc>;
+
+#[cfg(test)]
+use crate::schema::{device, realm_role, user_ as user};
+#[cfg(test)]
+use tests_fixtures::Device;
+
+#[derive(Debug, Queryable, Serialize, Deserialize, PartialEq)]
```
Comment (touilleMan) [inline oxidation/parsec_cli/parsec_sequestre_dump/src/model.rs:20]: Why serialize/deserialize ? those are supposed to represent data from the sqlite database and are never going to be serialized as-is

File: oxidation/parsec_cli/parsec_sequestre_dump/src/model.rs [hunk 13]
```diff
@@ -0,0 +1,519 @@
...
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Data")
+            .field("name", &self.name)
+            .field("children", &self.children)
+            .finish()
+    }
+}
+
+impl Default for Data {
+    fn default() -> Self {
```
Comment (touilleMan) [inline oxidation/parsec_cli/parsec_sequestre_dump/src/model.rs:73]: We never create Data object in this code (since it always code from the database)

I guess you only need this in the tests, if so would it be possible to instead define a `generate_empty_data()` function in the tests ?

File: oxidation/parsec_cli/parsec_sequestre_dump/src/model.rs [hunk 14]
```diff
@@ -0,0 +1,519 @@
...
+            _ => unreachable!(),
+        })
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct Workspace(pub Data);
+
+impl Workspace {
+    pub fn dump(&self, path: &str) -> Result<(), Box<dyn Error>> {
```
Comment (touilleMan) [inline oxidation/parsec_cli/parsec_sequestre_dump/src/model.rs:209]: Why box dyn error ? this seems overkill to me...

File: oxidation/parsec_cli/parsec_sequestre_dump/src/model.rs [hunk 15]
```diff
@@ -0,0 +1,519 @@
...
+        let data = serde_json::to_string_pretty(self)?;
+        std::fs::write(path, &data)?;
+
+        Ok(())
+    }
+
+    pub fn query_all(conn: &mut SqliteConnection, key: &SecretKey) -> Result<Self, Box<dyn Error>> {
+        let workspace_id = info::table.select(info::realm_id).first::<Vec<u8>>(conn)?;
```
Comment (touilleMan) [inline oxidation/parsec_cli/parsec_sequestre_dump/src/model.rs:217]: We should be using the magic value to retrieve the row (and add a comment the fact we do this to have a simple sanity check that brings a high level of confidence the file contain an sqlite db with valid data)

On top of that we should also check the version field.

File: oxidation/parsec_cli/parsec_sequestre_dump/src/model.rs [hunk 16]
```diff
@@ -0,0 +1,519 @@
...
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq)]
+pub struct Workspace(pub Data);
+
+impl Workspace {
+    pub fn dump(&self, path: &str) -> Result<(), Box<dyn Error>> {
+        let data = serde_json::to_string_pretty(self)?;
+        std::fs::write(path, &data)?;
```
Comment (touilleMan) [inline oxidation/parsec_cli/parsec_sequestre_dump/src/model.rs:211]: I guess this code is not finished: the cli is supposed to write all the files/folders from the workspace into the output folder

**Review by touilleMan — CHANGES_REQUESTED** [review]

**Comment by touilleMan** [comment]
For the tests, you'd better use the following pattern:
1 - init an empty sqlite db
2 - have a SQL command with plenty of INSERT inlined in your test that you run to put data in your db
3 - run you export command with a temporary folder as output
4 - check the output folder has the expect hierarchy and file contents

File: oxidation/crates/parsec_api_types/src/manifest.rs [hunk 17]
```diff
@@ -438,15 +422,21 @@ impl_transparent_data_format_conversion!(
...
 );
 
-/*
- * FileOrFolderManifest
- */
-
 #[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
-#[serde(tag = "type")]
-pub enum FileOrFolderManifest {
-    #[serde(rename = "file_manifest")]
```
[the remark this answers is not in the document]
Comment (TimeEngineer) [reply oxidation/crates/parsec_api_types/src/manifest.rs:448]: Dead code enum

File: oxidation/parsec_cli/parsec_sequestre_dump/src/main.rs [hunk 18]
```diff
@@ -0,0 +1,176 @@
...
+    let mut conn = SqliteConnection::establish(input)?;
+
+    let key_file = std::fs::read(key_path)?;
+    let key = SecretKey::try_from(&key_file[..])?;
+    let workspace = Workspace::query_all(&mut conn, &key)?;
+
+    workspace.dump(output)
+}
+
+#[cfg(test)]
```
[the remark this answers is not in the document]
Comment (TimeEngineer) [reply oxidation/parsec_cli/parsec_sequestre_dump/src/main.rs:49]: It's different, in the case of tests/foo.rs, it's an integration test, you don't have access to private data (module/fields)

File: oxidation/crates/parsec_api_types/tests/test_addr.rs (hunk 6 above)
Comment (TimeEngineer) [reply oxidation/crates/parsec_api_types/tests/test_addr.rs:161]: Clippy complains about that

File: oxidation/parsec_cli/parsec_sequestre_dump/src/model.rs (hunk 14 above)
Comment (TimeEngineer) [reply oxidation/parsec_cli/parsec_sequestre_dump/src/model.rs:209]: Because this is simply a CLI, do you want to handle errors or just throw them in stderr ?

File: oxidation/parsec_cli/parsec_sequestre_dump/Cargo.toml (hunk 3 above)
Comment (FirelightFlagboy) [reply oxidation/parsec_cli/parsec_sequestre_dump/Cargo.toml:1]: The description will be the same for the package and the `about` of the cli, 
so it would be simpler to declare it once in the manifest

File: oxidation/parsec_cli/parsec_sequestre_dump/src/main.rs (hunk 5 above)
Comment (TimeEngineer) [reply oxidation/parsec_cli/parsec_sequestre_dump/src/main.rs:61]: Done in #2218

File: oxidation/parsec_cli/parsec_sequestre_dump/src/main.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (TimeEngineer) [reply oxidation/parsec_cli/parsec_sequestre_dump/src/main.rs:49]: Moreover parsec_sequestre_dump is not a lib

File: oxidation/parsec_cli/parsec_sequestre_dump/Cargo.toml (hunk 3 above)
Comment (TimeEngineer) [reply oxidation/parsec_cli/parsec_sequestre_dump/Cargo.toml:1]: Do you have any suggestion about the message ? I don't know what to say :sweat_smile:

File: oxidation/parsec_cli/parsec_sequestre_dump/src/main.rs [hunk 19]
```diff
@@ -0,0 +1,176 @@
...
+
+    use crate::model::{Data, Workspace};
+
+    #[rstest]
+    fn test_parsec_sequestre_dump(alice: &Device) {
+        use diesel::connection::SimpleConnection;
+
+        let input = "/tmp/parsec_sequestre_dump.sqlite";
+        let output = "/tmp/out";
+        let dir = std::env::var("CARGO_MANIFEST_DIR").expect("CARGO_MANIFEST_DIR should be set");
```
[the remark this answers is not in the document]
Comment (TimeEngineer) [reply oxidation/parsec_cli/parsec_sequestre_dump/src/main.rs:63]: Need to know where are migrations

**Comment by TimeEngineer** [comment]
Waiting Storage's PR for some improvements:
- tmp_path

File: oxidation/parsec_cli/parsec_sequestre_dump/Cargo.toml (hunk 3 above)
Comment (FirelightFlagboy) [reply oxidation/parsec_cli/parsec_sequestre_dump/Cargo.toml:1]:
```
Extract encrypted files from a parsec export
```

**Review by FirelightFlagboy — CHANGES_REQUESTED** [review]
The `CLI` is missing help message

File: oxidation/parsec_cli/parsec_sequestre_dump/src/model.rs (hunk 13 above)
Comment (TimeEngineer) [reply oxidation/parsec_cli/parsec_sequestre_dump/src/model.rs]: No need anymore

**Comment by TimeEngineer** [comment]
> The `CLI` is missing help message

You are rude :(, we already have
```
parsec_sequestre_dump 0.0.0
Simple program that dumps files and folders in a file

USAGE:
    parsec_sequestre_dump --input <INPUT> --key <KEY> --output <OUTPUT>

OPTIONS:
    -h, --help               Print help information
    -i, --input <INPUT>      Sqlite input file (e.g `my/workspace_export.sqlite`)
    -k, --key <KEY>          Secret key (e.g `my/keyfile.key`)
    -o, --output <OUTPUT>    Output directory (e.g `./output`)
    -V, --version            Print version information
```

File: oxidation/parsec_cli/parsec_sequestre_dump/src/model.rs (hunk 12 above)
Comment (TimeEngineer) [reply oxidation/parsec_cli/parsec_sequestre_dump/src/model.rs]: Done :+1:

**Comment by FirelightFlagboy** [comment]
> You are rude :(, we already have

I've just learn something.

> that clap now handle doc-comment

File: oxidation/parsec_cli/parsec_sequestre_dump/src/model.rs [hunk 20]
```diff
@@ -0,0 +1,129 @@
...
+        manifest_id: &[u8],
+        devices: &mut HashMap<i32, (VerifyKey, DeviceID)>,
+    ) -> Result<(), Box<dyn Error>> {
+        let (manifest, author, timestamp) = vlob_atom::table
+            .select((vlob_atom::blob, vlob_atom::author, vlob_atom::timestamp))
+            .filter(vlob_atom::vlob_id.eq(manifest_id))
+            .order_by(vlob_atom::timestamp.desc())
+            .first::<(Vec<u8>, i32, f64)>(conn)?;
+
+        let (verify_key, device_id) = if let Some(device) = devices.get(&author) {
```
Comment (touilleMan) [inline oxidation/parsec_cli/parsec_sequestre_dump/src/model.rs:78]: Why not loading all devices at once before doing the dump ? the device table is very small so having it in memory is not an issue, and this simplify the code a bit

File: oxidation/parsec_cli/parsec_sequestre_dump/src/tests.rs [hunk 21]
```diff
@@ -0,0 +1,392 @@
...
+use tests_fixtures::{alice, tmp_path, Device, TmpPath};
+
+impl Workspace {
+    pub fn init_db(conn: &mut SqliteConnection, device: &Device) {
+        use std::collections::HashMap;
+
+        use parsec_api_types::DeviceCertificate;
+
+        let t0 = "2000-1-1T00:00:00Z".parse::<DateTime>().unwrap();
+        let _t0 = t0.get_f64_with_us_precision();
```
Comment (touilleMan) [inline oxidation/parsec_cli/parsec_sequestre_dump/src/tests.rs:25]: `t0_as_f64` ? (the leading underscore let it suggest it is not used)

File: oxidation/parsec_cli/parsec_sequestre_dump/src/model.rs [hunk 22]
```diff
@@ -0,0 +1,129 @@
...
+            devices.get(&author).expect("Inserted before")
+        };
+
+        match Manifest::decrypt_verify_and_load(
+            &manifest,
+            key,
+            verify_key,
+            device_id,
+            DateTime::from_f64_with_us_precision(timestamp),
+        )? {
```
Comment (touilleMan) [inline oxidation/parsec_cli/parsec_sequestre_dump/src/model.rs:98]: I think we should be much more permissive on error handling: Currently if we encounter an invalid manifest we stop with an error.
Instead we should print ` Export error: invalid data (EntryID: 123, Path: /foo/bar)` (with `invalid data`/`bad signature`/`bad encryption`) and carry on.

**Review by touilleMan — APPROVED** [review]
